src/src/Controller/AbandonedCartController.php

```diff
@@ -0,0 +1,241 @@
+
+        if (!empty($subdomain) && !empty($username) && !empty($password)) {
+            $this->api = new Api($subdomain, $username, $password);
+        }
+    }
+
+    public function sync(string $token): bool
+    {
+        if ($this->configuration->get('SMAILY_CART_CRON_TOKEN') !== $token) {
+            echo 'Access denied!';
+
+            return false;
+        }
+
+        if (!$this->configuration->getBoolean('SMAILY_ENABLE_ABANDONED_CART')) {
+            echo 'Abandoned cart disabled!';
+
+            return false;
+        }
+
+        if (empty($this->api)) {
+            echo 'Account setup not finished!';
+
+            return false;
+        }
+
+        $count = 0;
+        $failCount = 0;
+
+        $abandonedCartsCollection = new AbandonedCartCollection();
+        foreach ($abandonedCartsCollection->carts as $cart) {
+            // Notice!
+            // When running cron job in the same browser the session is shared
+            // and the abandoned cart `date_upd` time is renewed on each request.
+            // This doesn't allow the check to pass. I recommend running the cron job
+            // in a private window to avoid this.
+            if (!$this->isDelayTimePassed($cart)) {
+                continue;
+            }
+
+            $this->send($cart) || $failCount++;
+
+            ++$count;
+        }
+
+        echo sprintf("%s abandoned cart email(s) sent!\r\n", $count);
+        if ($failCount > 0) {
+            echo sprintf("%s failed cart(s), check logs!\r\n", $failCount);
+        }
+
+        return $failCount > 0;
+    }
+
+    private function isDelayTimePassed(AbandonedCart $cart): bool
+    {
+        $cart_updated_time = strtotime($cart->dateUpdated);
+        $syncInterval = $this->configuration->get('SMAILY_ABANDONED_CART_TIME');
+
+        $reminder_time = strtotime('+' . $syncInterval . ' minutes', $cart_updated_time);
+        $current_time = strtotime(date('Y-m-d H:i') . ':00');
+
+        // Don't continue if cart delay time has not passed.
+        return $current_time >= $reminder_time;
+    }
+
+    /**
+     * Send abandoned cart emails to customers.
+     *
+     * @return void
+     */
+    private function send(AbandonedCart $cart): bool
+    {
+        $response = $this->api->triggerAutomation(
+            $this->configuration->get('SMAILY_CART_AUTORESPONDER'),
+            [$this->generatePayload($cart)],
+        );
+
+        if ($response->getStatusCode() !== 200) {
+            Logger::logErrorWithFormatting(
+                'Failed sending out abandoned cart email for email: %s, cart_id: %s. ' .
+                'Smaily response HTTP response code: %s.',
+                $cart->email,
+                $cart->cartID,
+                $response->getStatusCode()
+            );
+
+            return false;
+        }
+
+        $body = json_decode($response->getBody()->getContents(), true);
+        if (!isset($body['code']) || $body['code'] !== 101) {
```

Comment:
Good point, might we consider this as a feature improvement?

Reply:
Maybe we should store the last response status code if we are improving the logic. Then it allows us to select which ones we might try again or skip. For example, general HTTP `504` responses should be tried again, malformed content with a smaily `203` response should not. This also means that each sent cart is added to the table regardless of the status and the `SELECT` statement should also be changed. I think this feature can also be done with an upgrade script that modifies the table by adding a column and changing the query to target specific cart statuses.

Reply:
A normalized error would work better with the solution you are proposing, because it creates better boundaries between functionalities, i.e. every query does not need to know what "HTTP 200 and code 203" means.

I am marking this topic resolved, because there is an upgrade path in the form of migrations.